src/pathfinding_igvc/src/AStar.cpp

```diff
@@ -6,10 +6,224 @@
  */
 
 #include <AStar.h>
+using namespace std;
+
+AStar::AStar(nav_msgs::OccupancyGrid occupancy_grid, GridPoint start,
+        GridPoint goal) {
+    this->_num_cols = occupancy_grid.info.width;
+    this->_num_rows = occupancy_grid.info.height;
+    this->_goal = goal;
+    this->_start = start;
+    this->_grid = occupancy_grid.data;
+
+    this->_cell_details = new CellDetail*[this->_num_rows];
+    for (int i = 0; i < this->_num_rows; i++) {
+        this->_cell_details[i] = new CellDetail[this->_num_cols];
+    }
+
+    /*
+     Create an open list
+     */
+    this->_open_list = set<GridPointWithScore>();
+
+    // Create a closed list and initialise it to false which means
+    // that no cell has been included yet
+    // This closed list is implemented as a boolean 2D array
+    this->_closed_list = new bool*[this->_num_rows];
+    for (int i = 0; i < this->_num_rows; i++) {
+        this->_closed_list[i] = new bool[this->_num_cols];
+    }
+}
 
 std::stack<AStar::GridPoint> AStar::run(nav_msgs::OccupancyGrid occupancy_grid,
                                         GridPoint start,
                                         GridPoint goal) {
-    // TODO Part 2
+    return AStar(occupancy_grid, start, goal).aStarSearch();
+}
+
+std::stack<AStar::GridPoint> AStar::aStarSearch() {
+    GridPoint start = this->_start;
+
+    // Initialising the parameters of the starting node
+    this->_cell_details[start.row][start.col].f = 0.0;
+    this->_cell_details[start.row][start.col].g = 0.0;
+    this->_cell_details[start.row][start.col].h = 0.0;
+    this->_cell_details[start.row][start.col].parent = start;
+
+    this->_open_list.insert(std::make_pair(0.0, std::make_pair(start.row, start.col)));
+
+    while (!this->_open_list.empty()) {
+        GridPointWithScore p = *this->_open_list.begin();
+
+        // Remove this vertex from the open list
+        this->_open_list.erase(this->_open_list.begin());
+
+        // Add this vertex to the open list
+        int row = p.second.first;
+        int col = p.second.second;
+        this->_closed_list[row][col] = true;
+
+        GridPoint parent(row, col);
+        /*
+        Generating all the 8 successor of this cell
+
+            N.W   N   N.E
+              \   |   /
+               \  |  /
+            W----CellDetail----E
+                 / | \
+               /   |  \
+            S.W    S   S.E
```

Comment:
Loving the ASCII art :+1: , but the center section looks like it's a few spaces to the right, and the bottom section looks like it's one space to the right.

Reply:
btw I didn't make this... LOOL it was taken from geeksforgeeks.com lmao